pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go

```diff
@@ -390,8 +411,29 @@ func (lrw *logicalReplicationWriterProcessor) handleEvent(
 	return nil
 }
 
+func (lrw *logicalReplicationWriterProcessor) maybeCheckpoint(
+	ctx context.Context, resolvedSpans []jobspb.ResolvedSpan,
+) error {
+
+	// If purgatory is empty, just emit the checkpoint now.
+	if len(lrw.purgatory) == 0 {
+		return lrw.checkpoint(ctx, resolvedSpans)
+	}
+
+	// If the current purgatory level is already closed, make a new one.
```

Comment:
what if we created a standalone purgatory object that contained `purgatoryEntries`? This purgatory object could have the following methods, which I think would create a nice abstraction:

```
purgatory.Add()
purgatory.IsEmpty()
purgatory.MaybeDrain()
purgatory.IsFull()
```

Reply:
I played with that initially, but it doesn't work out as all that standalone so I moved away from it:

You'd ideally want a one-way dependency, but draining is an action LRW takes, not the purgatory, so it ends up just being a read method. If all we're doing is appending, checking length, and reading, then it is hard to see what it is other than a slice?

Reply:
hmmm, i see what you mean. What if `purgatory.MaybeDrain()` took in flushKVs and flushCheckpoint as args? This patch seems real hard to unit test without a bit more abstraction.

Reply:
tried this in latest revision.